modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/metadata/IgniteMdColumnOrigins.java

```diff
@@ -0,0 +1,362 @@
+        }
+        // Anything else is a derivation, possibly from multiple columns.
+        final Set<RelColumnOrigin> set = getMultipleColumns(rexNode, input, mq);
+
+        return createDerivedColumnOrigins(set);
+    }
+
+    /** */
+    public @Nullable Set<RelColumnOrigin> getColumnOrigins(Filter rel, RelMetadataQuery mq, int iOutputColumn) {
+        return mq.getColumnOrigins(rel.getInput(), iOutputColumn);
+    }
+
+    /** */
+    public @Nullable Set<RelColumnOrigin> getColumnOrigins(Sort rel, RelMetadataQuery mq, int iOutputColumn) {
+        return mq.getColumnOrigins(rel.getInput(), iOutputColumn);
+    }
+
+    /** */
+    public @Nullable Set<RelColumnOrigin> getColumnOrigins(TableModify rel, RelMetadataQuery mq, int iOutputColumn) {
+        return mq.getColumnOrigins(rel.getInput(), iOutputColumn);
+    }
+
+    /** */
+    public @Nullable Set<RelColumnOrigin> getColumnOrigins(Exchange rel, RelMetadataQuery mq, int iOutputColumn) {
+        return mq.getColumnOrigins(rel.getInput(), iOutputColumn);
+    }
+
+    /** */
+    public @Nullable Set<RelColumnOrigin> getColumnOrigins(
+        TableFunctionScan rel,
+        RelMetadataQuery mq,
+        int iOutputColumn
+    ) {
+        Set<RelColumnMapping> mappings = rel.getColumnMappings();
+
+        if (mappings == null) {
+            if (!rel.getInputs().isEmpty()) {
+                // This is a non-leaf transformation:  say we don't
+                // know about origins, because there are probably
+                // columns below.
+                return null;
+            }
+            else {
+                // This is a leaf transformation: say there are for sure no
+                // column origins.
+                return Collections.emptySet();
+            }
+        }
+
+        final Set<RelColumnOrigin> set = new HashSet<>();
+
+        for (RelColumnMapping mapping : mappings) {
+            if (mapping.iOutputColumn != iOutputColumn)
+                continue;
+
+            final RelNode input = rel.getInputs().get(mapping.iInputRel);
+            final int column = mapping.iInputColumn;
+            Set<RelColumnOrigin> origins = mq.getColumnOrigins(input, column);
+
+            if (origins == null)
+                return null;
+
+            if (mapping.derived)
+                origins = createDerivedColumnOrigins(origins);
+
+            set.addAll(origins);
+        }
+
+        return set;
+    }
+
+    /**
+     * Get column origins.
+     *
+     * @param rel Rel to get origins from.
+     * @param mq Rel metadata query.
+     * @param iOutputColumn Column idx.
+     * @return Set of column origins.
+     */
+    public @Nullable Set<RelColumnOrigin> getColumnOrigins(
+        ProjectableFilterableTableScan rel,
+        RelMetadataQuery mq,
+        int iOutputColumn
+    ) {
+        if (rel.projects() != null) {
+            if (rel.getInputs().isEmpty() || rel.getInputs().size() > 1)
```

Comment:
TableScan actually is a leaf node, so it has no inputs and it's OK

Reply:
Apparently, dummy copy-pasting is not an option here, so this whole if-branch should be reworked